File: Q763_PartitionLabels.py
class Solution(object):
    def partitionLabels(self, S):
	last = {c: i for i, c in enumerate(S)} #以字母为键，索引为值生成字典
	j = anchor = 0
	ans = [] #储存结果的列表
	for i, c in enumerate(S):
		j = max(j, last[c])	 #对每个位置的字母，若该位置已是最大位置则放入结果列表
		if i == j:
		ans.append(i - anchor + 1)
	    anchor = i + 1
																									        
	return ans
'''
分析
解决 ：划分标签问题
采用算法：采用了贪心算法，对每个位置的字符都对比是否是该字符的最大位置
Time：O(n)
Space：O(n)
TestRuntime：55ms
'''

"""对于注释双引号能够包含单引号
以下为自己的python解法：
class Solution:
    def partitionLabels(self, S):
        """
        :type S: str
        :rtype: List[int]
        """
        last = {}
        ans = []
        for i in range(len(S)): last[S[i]] = i
        length = last[S[0]]
        for j in range(len(S)): 
            if(length < last[S[j]]): length = last[S[j]]
            elif(j == length and last[S[j]] == length): 
                ans.append(length)
                if(j == len(S)-1): break
                length = last[S[j+1]] #该段报错字符串索引超出范围:取值时进行判断
        '''

        k = len(ans)
        while(k > 0):
            ans[k] = ans[k] - ans[k-1] #报错列表索引超出范围:不是0超出范围，而是len(ans)为长度，比位置大1超出了
            k = k - 1
            
        ans[0] = ans[0] + 1
        '''   
        for i in range(len(ans)-1,0,-1): ans[i] = ans[i] - ans[i-1]
        ans[0] = ans[0] + 1
        return ans
"""

File: Q763_PartitionLabels.cpp
class Solution {
public:
    vector<int> partitionLabels(string S) {
        //第一次遍历字符串并生成其中每个字符的最远位置然后用字典储存;
        //第二次遍历选择首个字符的最远位置为当前分部的长度，然后向后遍历若当前位置字符的最远位置小于长度则跳过;若大于修改长度;等于则为分段尾
        map<char,int> last;//cpp中的map模板库程初始化实例
        for(int i = 0; i < S.length(); i++){ //遍历并生成字典
            last[S[i]] = i; //此处存在一个非逻辑问题（语法）:初始化映射时键应当是char（单个字符）
        } 
        /*
        map<char, int> ::iterator it;//使用迭代器遍历map
        it = last.begin();
        while(it != last.end()){
            cout<<it->first;//输出键
            cout<<it->second<<endl;//输出值
            it++;
        }
        */
        vector<int> ans; //初始化结果矢量（列表）
        int length = last[0]; //将分部首个元素的长度设为分部长度，直接跳到下一个分部的起始元素;考虑从0开始，长度需要+1
        for(int i = 0; i < S.length(); i++){ //整部分进行遍历
            //对长度大于当前分部长度的字符进行扩展
            if(last[S[i]] > length) length = last[S[i]];
            else if((last[S[i]] == length)&&(i == length)){ //长度刚好等于分部长度并且该厂读就是最大长度位置，说明是当前分部的最后一个元素
                ans.insert(ans.end(), length); //此处有逻辑错误:插入位置不是数值，而是指针
                //另外此处需要考虑减去已分段的长度：另作处理，并且第一端长度需+1
                length = last[S[i+1]];
            }
        }
        for(int i=ans.size(); i>=0; i--) ans[i] = ans[i] - ans[i-1];
        ans[0] += 1;
        return ans;
    }
};

/*
别人的答案：思路相同，但是代码简略得多
vector<int> partitionLabels(string S) {
  vector<int> res, pos(26, 0);  直接用长度为26的vector充当字典
  for (auto i = 0; i < S.size(); ++i) pos[S[i] - 'a'] = i; 
  for (auto i = 0, idx = INT_MIN, last_i = 0; i < S.size(); ++i) {
    idx = max(idx, pos[S[i] - 'a']); 直接使用内置函数max
    if (idx == i) res.push_back(i - exchange(last_i, i + 1) + 1); vector的插入元素方法push_back
  }
  return res;
}
*/
